src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java

```diff
@@ -0,0 +1,279 @@
+        }
+
+        if(Objects.nonNull(paymentRequestRepository.findByIdTransaction(idTransaction))) {
+            log.warn("Transaction " + idTransaction + " has already been processed previously");
+            return createxPayAuthResponse(TRANSACTION_ALREADY_PROCESSED_MSG, HttpStatus.UNAUTHORIZED, null);
+        }
+        log.info("END POST - " + REQUEST_PAYMENTS_XPAY);
+        return createAuthPaymentXPay(pgsRequest, clientId, mdcFields);
+    }
+
+    @GetMapping(XPAY_AUTH)
+    public ResponseEntity<XPayAuthPollingResponse> getXPayAuthorizationResponse(@PathVariable String requestId,
+                                                                @RequestHeader(required = false, value = MDC_FIELDS) String mdcFields) {
+        log.info(("SART - get XPay Authorization response for requestId: " + requestId));
+        setMdcFields(mdcFields);
+
+        PaymentRequestEntity entity = paymentRequestRepository.findByGuid(requestId);
+        if(Objects.isNull(entity) || !entity.getRequestEndpoint().equals(REQUEST_PAYMENTS_XPAY)){
+            log.error("No xPay request entity has been found for requestId: " + requestId);
+            XPayPollingResponseError error = new XPayPollingResponseError(404L, NOT_FOUND_MSG);
+            return createXPayAuthPollingResponse(HttpStatus.NOT_FOUND, error, null);
+        }
+        return  createXPayAuthPollingResponse(HttpStatus.OK, null, entity);
+    }
+
+    private ResponseEntity<XPayAuthResponse> createxPayAuthResponse(String errorMessage, HttpStatus status, String requestId) {
+        XPayAuthResponse response = new XPayAuthResponse();
+        if(Objects.nonNull(requestId)) {
+            PaymentRequestEntity entity = paymentRequestRepository.findByGuid(requestId);
+            response.setRequestId(requestId);
+            response.setStatus(entity.getStatus());
+        }
+        if(StringUtils.isEmpty(errorMessage)) {
+            String urlRedirect = StringUtils.join(PGS_RESPONSE_URL_REDIRECT, requestId);
+            response.setUrlRedirect(urlRedirect);
+        } else {
+            response.setError(errorMessage);
+            log.info("END POST - " + REQUEST_PAYMENTS_XPAY);
+        }
+        return ResponseEntity.status(status).body(response);
+    }
+
+    private ResponseEntity<XPayAuthResponse> createAuthPaymentXPay(XPayAuthRequest pgsRequest, String clientId, String mdcFields) {
+       log.info("START - XPay Request Payment Authorization ");
+
+        PaymentRequestEntity paymentRequestEntity = new PaymentRequestEntity();
+        String transactionId = pgsRequest.getIdTransaction();
+        try{
+            AuthPaymentXPayRequest xPayRequest = createXPayRequest(pgsRequest);
+            String authRequestJson = OBJECT_MAPPER.writeValueAsString(xPayRequest);
+            generateRequestEntity(clientId, mdcFields, transactionId, paymentRequestEntity, xPayRequest.getTimeStamp());
+            paymentRequestEntity.setJsonRequest(authRequestJson);
```

Comment:
this line can be moved in the "generateRequestEntity" method since it sets an entity field